### PR TITLE
install

### PR DESCRIPTION
**Pull request by jburel**
This PR:
 * Install fiji BF and OMERO.ij plugin
 * Install xvfb so BF can be used to open images
 * Modify start up script so Fiji/BF can be used in notebook
 * Turn groovy crop script from training-script into a notebook
 * install orbit
 * Convert orbit script into a notebook
 * install ilastik
 * add pixels classification example for ilastik

To test:
* build the image locally ``docker build -t training-notebooks .``
* start the image ``docker run -it  -p 8888:8888 training-notebooks``
* Run the notebooks under Fiji/Orbit/ilastik

cc @pwalczysko

### CODE REVIEW

**Comment by pwalczysko** [comment]
Screen Shot 2019-09-26 at 15 02 42

I am not sure at this stage whether or not I am logged in. 
I have put in the correct creds and id, but when hitting Enter, nothing appears.
If I re-execute the cell, then I am asked to fill in the form once again.

**Comment by pwalczysko** [comment]
The workflow went through successfully, the crop was imported. 

What I was looking for after that was how to put in another image id.
I had to scroll up to the previous cell, which worked okay - but maybe there could be a popup asking me about the image ID only after I have clicked onto the cell with the code itself ? This means, do not bind the image ID input to the creds cell, instead, give the image id more flexibility.

2 comments, Otherwise it works nicely.

**Comment by jburel** [comment]
cf. collection of credentials it does not move automatically to the next cell.
Hitting enter does not log you in
logging happens in the next cell

Same problem that the one we have with the other notebook
I will split the connection and ID

**Comment by dominikl** [comment]
The forms a quite confusing, not really clear what's going on. Why not just let user put in username/password etc. in the notebook itself? For orbit, don't you have to select a small area to run the segmentation on? Can't see where this happens in the code. Bit scared to kick off the segmentation and run it over the whole 60k x 60k pixels image.

**Comment by jburel** [comment]
Orbit: this is the same as the script we run in orbit script editor. Nothing new
For the form, people want to have things hardcoded especially when they want to try against their own data
What do you find confusing in the form? It is a standard login form

**Comment by dominikl** [comment]
Ok, somehow Orbit automatically restricts itself to only segment the ROI of the image. Re forms: Alright. Maybe for the user the forms are actually more intuitive than changing variables in the code. For me it just wasn't clear if I had to re-run the cell after putting something in the form, etc. Generally 👍

**Comment by pwalczysko** [comment]
> Maybe for the user the forms are actually more intuitive than changing variables in the code. 

There is one additional strong reason for the "forms" - password not in the open.

**Comment by pwalczysko** [comment]
I think the form becomes much less confusing if you get a clear message in along the lines "you are logged in now".

File: Fiji/crop_image.ipynb [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+    "\n",
+    "\n",
+    "This notebook shows how to crop an OMERO image using the crop functionality available in [Fiji](https://imagej.net/Fiji). Fiji has been installed with few other plugins. For more information, check the Dockerfile."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First we had Fiji and the plugins to the classpath."
```
Comment (mtbc) [inline Fiji/crop_image.ipynb:17]:
```suggestion
    "First we add Fiji and the plugins to the classpath."
```
perhaps?

**Comment by jburel** [comment]
The form only collects the credentials, the connection happens much later in the main script.

**Comment by jburel** [comment]
@pwalczysko @dominikl I have re-organised a bit the notebook so the connection happens after collecting the credentials

**Comment by pwalczysko** [comment]
Ilastik: The probabilities images coming from ilastik are clearly flipped in comparison with the original.

Screen Shot 2019-10-02 at 16 03 40

I think it is a flip along the midline which goes from left to right accross the image - what is top on the original, is bottom on the ilastik probabilities

**Comment by pwalczysko** [comment]
cc @will-moore

**Comment by jburel** [comment]
I will check. I thought I sorted it

**Comment by jburel** [comment]
@pwalczysko was not using the latest version.

**Comment by pwalczysko** [comment]
Screen Shot 2019-10-03 at 10 37 40


the error is specific for the ilastik notebook, other notebooks are loading.

**Comment by jburel** [comment]
@pwalczysko fixing it now

**Comment by pwalczysko** [comment]
Screen Shot 2019-10-03 at 10 56 37

I would suggest:

``Enter credentials to connect to OMERO`` -> ``Enter credentials and continue to the next cell. Do not re-run this cell.`` Otherwise I really do not know what to do.

**Comment by pwalczysko** [comment]
Screen Shot 2019-10-03 at 11 00 16

Could the  ``null`` message be supressed somehow ? Makes me think that from my connection attempt, the result is ``null`` -> not connected

**Comment by jburel** [comment]
@pwalczysko I have not found a way to do that yet

**Comment by jburel** [comment]
@pwalczysko ilastik notebook should now be fixed

**Comment by pwalczysko** [comment]
the ilastik notebook works nicely now

**Comment by pwalczysko** [comment]
trainer-1 on outreach has macro attached on ``A-Fiji-dataset`` - id 2920. Macro name ``idr0021-segment.ijm``

**Comment by jburel** [comment]
@pwalczysko saving the output should now be in place
In order to be as "generic" as possible
The should add something like that  to the macrofile:
//saveImageToOMERO
//saveROIToOMERO
//saveResultsToOMERO
the notebook should then parse the file check its contain and see what are the saving options
e.g. for a macro like "crop", user will have to add 
```
//saveImageToOMERO
```
for the macro currently attached
```
//saveROIToOMERO
//saveResultsToOMERO
```
i.e. user will have to edit the macro before uploading it

**Comment by pwalczysko** [comment]
I have added the two last lines to the macro stored on outreach trainer-1 A-Fiji-dataset.
Then I ran the notebook.

It does not seem to work - I am not getting any other output for ages than

```
loading images...
loading macro...
analyzing images...
Running macro on image: 9542
```

**Comment by jburel** [comment]
The options in the macro i.e. ``//saveROIToOMERO`` are just a suggestion.
The notebook should save the data

edit: something is blocking the notebook. Need to check

**Comment by pwalczysko** [comment]
```
loading images...
loading macro...
analyzing images...
Image opened: 55459
Running macro crop100.ijm on image: 55459
importing...
Image:55461

Other imported objects:
Fileset:39332

imported
Image opened: 55458
Running macro crop100.ijm on image: 55458

java.nio.file.FileAlreadyExistsException: /tmp/Fiji_csv17344283800208997356/Centrin_PCNT_Cep215_20110506_Fri-1620_0_SIR_PRJ.ome.tif_results.csv
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at script1571327094836$_run_closure1.doCall(script1571327094836:256)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at this cell line 234
```

Edit: I think this error is there only when the crop macro is being reused twice on the same dataset. I guess it is a fail on re-importing the image with the same name twice  ? Athough not sure why this is a problem...

**Comment by pwalczysko** [comment]
The segmentation principally works, but, it delivers nonsensical ROIs for this type of images.

Both with MaxEntropy and Default.

I think we need to use MaxEntropy and then invert the image - will try.

**Comment by pwalczysko** [comment]
The segmentation macro works pretty nicely on the "condensation" images

Screen Shot 2019-10-17 at 17 10 35

**Comment by jburel** [comment]
The accuracy of the macro is not really related to this PR.
It is just a way to generate output that we can handle
The commit  https://github.com/ome/training-notebooks/pull/40/commits/cb1e9e0ee6e358d52e8df8d5f6a8068bd12a26b5 should fix https://github.com/ome/training-notebooks/pull/40#issuecomment-543237616

**Comment by pwalczysko** [comment]
```
java.nio.file.FileAlreadyExistsException: /tmp/Fiji_csv13545928975078592064/A1.ome.tif_results.csv
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

run the crop.ijm, just selecting Images to save on the Dataset "test-jm"

**Comment by pwalczysko** [comment]
The last fix works. 

I think this is fine to merge.

Will have to find a way how to produce white object on black bcgr though.

**Comment by jburel** [comment]
travis started to fail, job now exceeds the time limit

**Comment by pwalczysko** [comment]
so, now, the macro works also on the "A-Fiji-dataset", see dataset 2306 of user-1 (Francis Crick).

@jburel : The "strange behaviour" we noticed in the office on that dataset was caused by the fact that the macro "Test-1.ijm" was belonging to trainer-1, so user-1 could not see it (although he should actually, because it is a RA group  .... He sees it in the UI). In the moment when I removed the "Test-1.ijim" and re-uploaded it as user-1, things started to work.


Screen Shot 2019-10-18 at 15 34 52

**Comment by pwalczysko** [comment]
I think the problem we are having here still is the permissions - what if the macro attached belongs to another user - then the owner of the dataset cannot see it via this notebook, but he can see it in the Webclient...

**Comment by pwalczysko** [comment]
But, there is no problem with the workflow and ROI quality and the CSVs when everything is run and owned by the same person. 

The macro

```
run("8-bit");
run("Invert", "stack");
setAutoThreshold("MaxEntropy");
run("Analyze Particles...", "size=10-Infinity pixel display clear add stack");
```

**Comment by jburel** [comment]
@pwalczysko so we now have the macro that we can use in ITR

**Comment by jburel** [comment]
@pwalczysko ok the permissions issue is not really an issue it is b/c
I only retrieve the annotations for a given user i.e.
```
users = new ArrayList(1)
users.add(new Long(user.getId()))
...
annotations = svc.getAnnotations(ctx, dataset, types, users)
```
So that can be easily adjusted

**Comment by pwalczysko** [comment]
Ok, Ready to merge then.

**Comment by pwalczysko** [comment]
Well, the travis is not happy though.

**Comment by jburel** [comment]
That's what I mentioned during the call. We now hit the time limit for job in Travis

**Comment by pwalczysko** [comment]
So, commenting out the line

```
//users.add(new Long(user.getId()))
```

caused that now also attachments of other users on that dataset are seen and the whole notebook still works, thanks @will-moore 
cc @jburel

**Comment by pwalczysko** [comment]
Last commit works as expected. I can run the Fiji notebook on a macro belonging to another user without any adjustments.

I think we could merge this.

I will need to adjust  https://omero-guides.readthedocs.io/en/latest/fiji/docs/index.html for the workshop

**Comment by pwalczysko** [comment]
I guess we should not hardcode the "outreach" server anymore anywhere. In the next outreach, the "workshop" server will be used.

**Comment by jburel** [comment]
I will open another PR to sort that out for the other notebooks

**Comment by pwalczysko** [comment]
It seems to me that there is a following bug still present in the Fiji (attached macro) workflow:

- have a dataset with 4 images
- have a working macro attached to it
- verify in the webclient how many ROIs do you have on each image prior to the run of the script
- connect jto the server (workshop, the trainer-1's "A-Fiji-dataset")
- run the notebook on that dataset, selecting to save both ROIs and Results
- observe that the first image picked by the script is not having any ROIs saved on it - but the subsequent three images do have ROIs as expected -this problem is also capture inside the notebook output itself, see the snippet below
- the csv is attached to all four images as expected


Screen Shot 2019-10-29 at 11 50 46


```
loading images...
loading macro...
analyzing images...
Image opened: 9542
Running macro segmentation.ijm on image: 9542
 	Area	Mean	Min	Max
1	0.052	174.970	107	219
2	0.024	176.800	124	217
3	0.072	158.109	0	219
4	0.016	197.700	170	219
5	0.019	206.333	194	220
6	0.049	198.968	152	220
7	0.017	201.818	183	217
8	0.577	158.220	0	220
9	0.019	209.917	198	219
10	0.016	207.700	193	220
11	0.019	212.917	201	219
12	0.016	208.700	194	220
13	0.020	208.077	187	220
14	0.016	206.700	193	217
15	0.028	188.944	132	220
16	0.020	209.385	191	220
17	0.016	204.900	188	219
18	0.024	213.333	200	220
19	0.017	204.636	185	219
20	0.016	216.300	206	220
21	0.075	200.667	155	220
22	0.061	202.615	159	220
23	0.045	184.310	146	220
24	0.045	182.241	131	220
25	0.024	200.533	173	220
26	0.017	198.455	175	214
27	0.016	212.500	206	220
28	0.016	207.300	194	217
29	0.030	209.105	198	220
30	0.033	208.190	193	220
31	0.022	203.071	178	219
32	0.990	170.705	0	220
33	0.016	211.800	206	217
34	0.019	208.417	190	220
35	0.025	184.500	137	214
36	0.030	181.000	115	217
37	0.016	209.000	195	218
38	0.064	209.390	194	220
39	0.017	207.727	194	219
40	0.022	210.429	197	220
41	0.020	201.077	175	220
42	0.016	198.600	176	220
43	0.016	203.000	184	220
44	0.016	197.300	173	216
45	0.024	191.067	165	220
46	0.016	208.800	195	218
47	0.027	207.294	196	219
48	0.016	211.600	202	219
saving Results as CSV
Image opened: 9540
Running macro segmentation.ijm on image: 9540
 	Area	Mean	Min	Max
1	0.049	131.032	7	191
2	0.075	140.458	30	195
3	0.034	127.045	20	195
4	0.060	131.842	0	192
5	0.083	165.245	109	195
6	0.016	174.500	147	193
7	0.241	151.260	0	195
8	0.017	173.727	155	192
9	0.017	145.636	92	195
10	0.016	177.200	157	193
11	0.583	148.691	0	195
saving ROIs
saving Results as CSV
Image opened: 9539
Running macro segmentation.ijm on image: 9539
 	Area	Mean	Min	Max
1	0.154	144.031	22	188
2	0.016	175.300	162	186
3	0.050	153.188	95	189
4	0.121	126.701	0	189
5	0.036	164.870	128	189
6	0.240	121.431	0	189
7	0.176	142.321	15	189
8	0.017	180.364	171	189
9	0.031	137.600	57	189
10	0.019	174.250	153	189
11	0.019	167.917	144	187
12	0.025	168.500	148	188
13	0.017	167.091	142	187
14	0.315	143.085	0	189
15	0.198	144.730	49	189
16	0.049	169.645	147	187
17	0.022	164.500	140	183
saving ROIs
saving Results as CSV
Image opened: 9541
Running macro segmentation.ijm on image: 9541
 	Area	Mean	Min	Max
1	0.028	136.333	54	204
2	0.045	119.655	0	201
3	0.019	191.083	173	203
4	0.024	139.600	58	205
5	0.019	162.667	106	205
6	0.196	124.400	15	205
7	0.205	148.206	0	204
8	0.016	173.000	141	203
9	0.028	178.278	131	205
10	0.016	190.800	171	205
11	0.027	149.412	75	205
12	0.213	164.735	0	205
13	0.364	147.578	23	205
saving ROIs
saving Results as CSV
processing done
```

**Pull request merged**